datascience/maps.py

```diff
@@ -0,0 +1,510 @@
+        Called before to_html, add points and regions before outputting map 
+        """
+        for point in self.points:
+            point.map(self._folium)
+
+        for region in self.regions:
+            region.map(self._folium)
+
+        # uncomment the following with new Folium release
+        # if 'zoom_start' not in self._init:
+        #     self._autofit(self.bounds)
+    
+    def _before_map(self, *args, **kwargs):
+        """ Autozoom map where appropriate """
+        self.bounds = self._autobounds()
+        
+        if not self.location:
+            self._autocenter(self.bounds)
+
+        # remove with new Folium release
+        if 'zoom_start' not in self._init:
+            self._autofit(self.bounds)
+            
+    def _autobounds(self):
+        """ Simple calculation for bounds """
+        bounds = {}
+
+        def check(prop, compare, extreme, val):
+            opp = min if compare is max else max
+            bounds.setdefault(prop, val)
+            bounds[prop] = opp(compare(bounds[prop], val), extreme)
+
+        def bound_check(coord):
+            if not coord:
+                return
+            if isinstance(coord, self._allowed_collections):
+                if isinstance(coord[0], (int, float)):
+                    long, lat = coord
+                    check('max_lat', max, 90, lat)
+                    check('min_lat', min, -90, lat)
+                    check('max_long', max, 180, long)
+                    check('min_long', min, -180, long)
+                else:
+                    [bound_check(c) for c in coord]
+            else:
+                raise ProgrammerError('Ended up with %s.' % type(coord))
+
+        bound_check([point.location for point in self.points])
+        bound_check([region.to_polygon() for region in self.regions])
+        return bounds
+        
+    def _autofit(self, bounds):
+        """ Automatically fits everything with the maximum zoom possible """
+        if not bounds:
+            return
+        
+        # uncomment with new Folium release
+        # self._folium.fit_bounds(
+        #     [bounds['min_long'], bounds['min_lat']],
+        #     [bounds['max_long'], bounds['max_lat']]
+        # )
+        
+        # remove the following with new Folium release
+        # rough approximation, assuming max_zoom is 18
+        import math
+        try:
+            factor = 1.2
+            width = ((bounds['max_lat'] - bounds['min_lat']) or 2)
+            height = ((bounds['max_long'] - bounds['min_long']) or 2)
+            area, max_area = width*height, 180*360
+            zoom = math.log(area/max_area)/-factor
+            self['zoom_start'] = max(1, min(18, round(zoom)))
+        except ValueError:
+            raise MapError('Uh oh. Is the data formatted as long-lat \
+            pairs? (Check. If the smallest number is less than -90, \
+            those are long-lat pairs) If so, please add geo_formatted=True \
+            to create_map')
```

Comment:
Cool.  Seems to work at reasonable scale.
On Jul 25, 2015 10:37 AM, "Alvin Wan" <notifications@github.com> wrote:

> In datascience/maps.py
> <https://github.com/dsten/datascience/pull/19#discussion_r35485001>:
>
> > +
> > +        # remove the following with new Folium release
> > +        # rough approximation, assuming max_zoom is 18
> > +        import math
> > +        try:
> > +            factor = 1.2
> > +            width = ((bounds['max_lat'] - bounds['min_lat']) or 2)
> > +            height = ((bounds['max_long'] - bounds['min_long']) or 2)
> > +            area, max_area = width*height, 180*360
> > +            zoom = math.log(area/max_area)/-factor
> > +            self['zoom_start'] = max(1, min(18, round(zoom)))
> > +        except ValueError:
> > +            raise MapError('Uh oh. Is the data formatted as long-lat \
> > +            pairs? (Check. If the smallest number is less than -90, \
> > +            those are long-lat pairs) If so, please add geo_formatted=True \
> > +            to create_map')
>
> :D Thanks
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/dsten/datascience/pull/19/files#r35485001>.
>